test-support/reference-trie/src/lib.rs

```diff
@@ -15,32 +15,129 @@
 //! Reference implementation of a streamer.
 
 use std::fmt;
-use std::error::Error as StdError;
 use std::iter::once;
+use std::marker::PhantomData;
 use parity_scale_codec::{Decode, Input, Output, Encode, Compact, Error as CodecError};
 use trie_root::Hasher;
-use trie_db::{node::Node, triedbmut::ChildReference, DBValue};
+use trie_db::{
+	node::Node,
+	triedbmut::ChildReference,
+	DBValue,
+	trie_visit,
+	TrieBuilder,
+	TrieRoot,
+	Partial,
+};
+use std::borrow::Borrow;
 use keccak_hasher::KeccakHasher;
 
-pub use trie_db::{Trie, TrieMut, NibbleSlice, NodeCodec, Recorder, Record};
+pub use trie_db::{Trie, TrieMut, NibbleSlice, Recorder, NodeCodec};
+pub use trie_db::{Record, TrieLayout, TrieConfiguration, nibble_ops};
 pub use trie_root::TrieStream;
 
-pub type RefTrieDB<'a> = trie_db::TrieDB<'a, keccak_hasher::KeccakHasher, ReferenceNodeCodec>;
-pub type RefTrieDBMut<'a> = trie_db::TrieDBMut<'a, KeccakHasher, ReferenceNodeCodec>;
-pub type RefFatDB<'a> = trie_db::FatDB<'a, KeccakHasher, ReferenceNodeCodec>;
-pub type RefFatDBMut<'a> = trie_db::FatDBMut<'a, KeccakHasher, ReferenceNodeCodec>;
-pub type RefSecTrieDB<'a> = trie_db::SecTrieDB<'a, KeccakHasher, ReferenceNodeCodec>;
-pub type RefSecTrieDBMut<'a> = trie_db::SecTrieDBMut<'a, KeccakHasher, ReferenceNodeCodec>;
-pub type RefLookup<'a, Q> = trie_db::Lookup<'a, KeccakHasher, ReferenceNodeCodec, Q>;
+/// Trie layout using extension nodes.
+pub struct ExtensionLayout;
 
-pub fn ref_trie_root<I, A, B>(input: I) -> <KeccakHasher as Hasher>::Out where
+impl TrieLayout for ExtensionLayout {
+	const USE_EXTENSION: bool = true;
+	type Hash = keccak_hasher::KeccakHasher;
+	type Codec = ReferenceNodeCodec;
+}
+
+impl TrieConfiguration for ExtensionLayout { }
+
+/// Trie layout without extension nodes, allowing
+/// generic hasher.
+pub struct GenericNoExtensionLayout<H>(PhantomData<H>);
+
+impl<H: Hasher> TrieLayout for GenericNoExtensionLayout<H> {
+	const USE_EXTENSION: bool = false;
+	type Hash = H;
+	type Codec = ReferenceNodeCodecNoExt;
+}
+
+impl<H: Hasher> TrieConfiguration for GenericNoExtensionLayout<H> { }
+
+/// Trie layout without extension nodes.
+pub struct NoExtensionLayout;
```

Comment:
Could probably be `pub type NoExtensionLayout = GenericNoExtensionLayout<keccak_hasher::KeccakHasher>;`?